homeassistant/components/airvisual_pro/__init__.py

```diff
@@ -1,42 +1,73 @@
 """The AirVisual Pro integration."""
 from __future__ import annotations
 
+import asyncio
+from contextlib import suppress
+from dataclasses import dataclass
 from datetime import timedelta
 from typing import Any
 
 from pyairvisual import NodeSamba
-from pyairvisual.node import NodeProError
+from pyairvisual.node import NodeConnectionError, NodeProError
 
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_IP_ADDRESS, CONF_PASSWORD, Platform
-from homeassistant.core import HomeAssistant, callback
+from homeassistant.const import (
+    CONF_IP_ADDRESS,
+    CONF_PASSWORD,
+    EVENT_HOMEASSISTANT_STOP,
+    Platform,
+)
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
 from homeassistant.helpers.entity import DeviceInfo, EntityDescription
 from homeassistant.helpers.update_coordinator import (
     CoordinatorEntity,
     DataUpdateCoordinator,
     UpdateFailed,
 )
 
-from .const import DOMAIN, LOGGER
+from .const import CONF_DATA, DOMAIN, LOGGER
+
+CONF_RELOAD_TASK = "reload_task"
 
 PLATFORMS = [Platform.SENSOR]
 
 UPDATE_INTERVAL = timedelta(minutes=1)
 
 
+@dataclass
+class AirVisualProData:
+    """Define a data class."""
+
+    coordinator: DataUpdateCoordinator
+    node: NodeSamba
+
+
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up AirVisual Pro from a config entry."""
+    node = NodeSamba(entry.data[CONF_IP_ADDRESS], entry.data[CONF_PASSWORD])
+
+    try:
+        await node.async_connect()
+    except NodeProError as err:
+        raise ConfigEntryNotReady() from err
 
     async def async_get_data() -> dict[str, Any]:
         """Get data from the device."""
         try:
-            async with NodeSamba(
-                entry.data[CONF_IP_ADDRESS], entry.data[CONF_PASSWORD]
-            ) as node:
-                return await node.async_get_latest_measurements()
+            data = await node.async_get_latest_measurements()
+        except NodeConnectionError as err:
+            data = hass.data[DOMAIN][entry.entry_id]
+            if CONF_RELOAD_TASK not in data:
+                data[CONF_RELOAD_TASK] = hass.async_create_task(
+                    hass.config_entries.async_reload(entry.entry_id)
+                )
```

Comment:
I think you could store this as a local variable instead of in `data[CONF_RELOAD_TASK]` since when the local variable is destroyed you know its been reloaded